crates/pera-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -182,31 +289,39 @@ impl DWalletMPCSession {
     }
 
     /// Stores a message in the pending messages map.
-    /// The code stores every new message it receives for that session,
-    /// and when we reach the end of delivery,
-    /// we will advance the session if we have a threshold of messages.
-    fn store_message(&mut self, round: usize, message: &DWalletMPCMessage) -> DwalletMPCResult<()> {
-        let epoch_store = self.epoch_store()?;
-        let party_id = authority_name_to_party_id(&message.authority, &epoch_store)?;
-        if self.pending_messages[round].contains_key(&party_id) {
-            return Err(DwalletMPCError::MaliciousParties(vec![party_id]));
+    /// Every new message received for a session is stored.
+    /// When a threshold of messages is reached, the session advances.
+    fn store_message(&mut self, message: &DWalletMPCMessage) -> DwalletMPCResult<()> {
+        let source_party_id =
+            authority_name_to_party_id(&message.authority, &*self.epoch_store()?)?;
+
+        let msg_len = self.pending_messages.len();
+        match self.pending_messages.get_mut(message.round_number) {
+            Some(party_to_msg) => {
+                if party_to_msg.contains_key(&source_party_id) {
+                    // Duplicate.
+                    return Err(DwalletMPCError::MaliciousParties(vec![source_party_id]));
+                }
+                party_to_msg.insert(source_party_id, message.message.clone());
+            }
+            // If next round.
+            None if message.round_number == msg_len => {
+                let mut map = HashMap::new();
+                map.insert(source_party_id, message.message.clone());
+                self.pending_messages.push(map);
+            }
+            _ => {
+                // Unexpected round number; rounds should grow sequentially.
+                return Err(DwalletMPCError::MaliciousParties(vec![source_party_id]));
+            }
         }
-        self.pending_messages[round].insert(party_id, message.message.clone());
         Ok(())
     }
 
     /// Handles a message by either forwarding it to the session
     /// or ignoring it if the session is not active.
     pub(crate) fn handle_message(&mut self, message: &DWalletMPCMessage) -> DwalletMPCResult<()> {
```

Comment:
I think this function is useless now, just delete it and call the inner function directly